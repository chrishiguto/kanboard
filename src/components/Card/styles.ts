import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    background: ${theme.colors.white};
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xsmall};
  `}
`

type TopProps = {
  hasTag: boolean
}

const topModifiers = {
  hasTag: () => css`
    grid-template-areas: 'title type';
  `
}

export const Top = styled.div<TopProps>`
  ${({ theme, hasTag }) => css`
    display: grid;
    grid-template-areas:
      'tag type'
      'title .';
    grid-template-columns: 0.65fr 0.35fr;
    margin-bottom: ${theme.spacings.xsmall};
    word-break: break-word;
    grid-row-gap: calc(${theme.spacings.xsmall} / 2);

    ${!hasTag && topModifiers.hasTag()}
  `}
`

export const TagContainer = styled.div`
  grid-area: tag;
`

export const Tag = styled.span`
  ${({ theme }) => css`
    display: inline-flex;

    padding: calc(${theme.spacings.xsmall} / 4);
    border-radius: calc(${theme.border.radius} / 2);

    text-transform: uppercase;
    font-weight: ${theme.font.bold};
    font-size: calc(${theme.font.sizes.medium} - 0.5rem);
    color: #fa6e88;
    background: #fef1f5;
  `}
`

export const TypeContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  grid-area: type;
`

export const Type = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: center;

    height: 2.4rem;
    width: 2.4rem;

    padding: calc(${theme.spacings.xsmall} / 4)
      calc(${theme.spacings.xsmall} / 2);
    border-radius: calc(${theme.border.radius} / 2);
    background: #ebeef3;
  `}
`

export const Title = styled.h3`
  ${({ theme }) => css`
    color: #707990;

    font-weight: ${theme.font.bold};
    line-height: 2.4rem;
    font-size: ${theme.font.sizes.xsmall};
  `}
`

export const Bottom = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`

export const Date = styled.p`
  ${({ theme }) => css`
    color: #707990;

    font-size: calc(${theme.font.sizes.xsmall} - 0.1rem);
  `}
`

export const Responsible = styled.div`
  ${({ theme }) => css`
    display: flex;
    justify-content: flex-end;

    border-radius: ${theme.border.radius};

    > img {
      width: 2.4rem;
      height: 2.4rem;
    }
  `}
`
