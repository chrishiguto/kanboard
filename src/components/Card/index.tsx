import { Draggable, DraggableProvided } from 'react-beautiful-dnd'

import * as S from './styles'

export type CardProps = {
  id: number
  tag?: string
  type: string
  title: string
  date: string
  index: number
  responsible: {
    img: string
    name: string
  }
}

const Card = ({
  tag,
  type,
  title,
  date,
  responsible,
  id,
  index
}: CardProps) => (
  <Draggable draggableId={'task-' + id} index={index}>
    {(provided: DraggableProvided) => (
      <S.Wrapper
        {...provided.draggableProps}
        {...provided.dragHandleProps}
        ref={provided.innerRef}
      >
        <S.Top hasTag={!!tag}>
          {!!tag && (
            <S.TagContainer>
              <S.Tag>{tag}</S.Tag>
            </S.TagContainer>
          )}
          <S.TypeContainer>
            <S.Type>{type}</S.Type>
          </S.TypeContainer>
          <S.Title>{title}</S.Title>
        </S.Top>

        <S.Bottom>
          <S.Date>{date}</S.Date>
          <S.Responsible>
            <img src={responsible.img} alt={responsible.name} />
          </S.Responsible>
        </S.Bottom>
      </S.Wrapper>
    )}
  </Draggable>
)

export default Card
