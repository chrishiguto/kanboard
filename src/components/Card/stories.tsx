import { Story, Meta } from '@storybook/react/types-6-0'
import Card, { CardProps } from '.'

export default {
  title: 'Card',
  component: Card,
  parameters: {
    backgrounds: {
      default: 'dark'
    }
  }
} as Meta

export const Default: Story<CardProps> = (args) => (
  <div style={{ maxWidth: 320, margin: '0 auto' }}>
    <Card {...args} />
  </div>
)

Default.args = {
  tag: 'Needs review',
  type: 'T',
  title: 'Amending Noxious Weed Seed Rule'
}
