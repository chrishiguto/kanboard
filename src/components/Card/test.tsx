import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import Card from '.'

const props = {
  tag: 'Needs review',
  type: 'T',
  title: 'Amending Noxious Weed Seed Rule',
  date: 'Due Aug 31',
  responsible: {
    img: '/img/icon-192.png',
    name: 'Person name'
  }
}

describe('<Card />', () => {
  it('should render the card with tag', () => {
    const { container } = renderWithTheme(<Card {...props} />)

    expect(screen.getByText(/needs review/i)).toBeInTheDocument()
    expect(screen.getByText(/due aug 31/i)).toBeInTheDocument()
    expect(
      screen.getByRole('heading', { name: /amending noxious weed seed rule/i })
    ).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /person name/i })).toHaveAttribute(
      'src',
      props.responsible.img
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the card without tag', () => {
    renderWithTheme(<Card {...props} tag={undefined} />)

    expect(screen.queryByText(/needs review/i)).not.toBeInTheDocument()
    expect(screen.getByText(/due aug 31/i)).toBeInTheDocument()
    expect(
      screen.getByRole('heading', { name: /amending noxious weed seed rule/i })
    ).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /person name/i })).toHaveAttribute(
      'src',
      props.responsible.img
    )
  })
})
