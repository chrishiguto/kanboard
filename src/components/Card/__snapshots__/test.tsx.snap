// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render the card with tag 1`] = `
.c0 {
  background: #FAFAFA;
  border-radius: 0.8rem;
  padding: 1.6rem;
}

.c1 {
  display: grid;
  grid-template-areas: 'tag type' 'title .';
  grid-template-columns: 0.65fr 0.35fr;
  margin-bottom: 1.6rem;
  word-break: break-word;
  grid-row-gap: calc(1.6rem / 2);
}

.c2 {
  grid-area: tag;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: calc(1.6rem / 4);
  border-radius: calc(0.8rem / 2);
  text-transform: uppercase;
  font-weight: 500;
  font-size: calc(1.6rem - 0.5rem);
  color: #fa6e88;
  background: #fef1f5;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  grid-area: type;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 2.4rem;
  width: 2.4rem;
  padding: calc(1.6rem / 4) calc(1.6rem / 2);
  border-radius: calc(0.8rem / 2);
  background: #ebeef3;
}

.c6 {
  color: #707990;
  font-weight: 500;
  line-height: 2.4rem;
  font-size: 1.2rem;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 {
  color: #707990;
  font-size: calc(1.2rem - 0.1rem);
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  border-radius: 0.8rem;
}

.c9 > img {
  width: 2.4rem;
  height: 2.4rem;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <span
        class="c3"
      >
        Needs review
      </span>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        T
      </div>
    </div>
    <h3
      class="c6"
    >
      Amending Noxious Weed Seed Rule
    </h3>
  </div>
  <div
    class="c7"
  >
    <p
      class="c8"
    >
      Due Aug 31
    </p>
    <div
      class="c9"
    >
      <img
        alt="Person name"
        src="/img/icon-192.png"
      />
    </div>
  </div>
</div>
`;
